### PR TITLE
Support cuDNN CTC functions

### PR DESCRIPTION
**Pull request by aonotas**
Thank you for creating a pull request!

cupy PR: https://github.com/cupy/cupy/pull/1769

Please double-check the following.

- Read [our contribution guide](https://docs.chainer.org/en/stable/contribution.html).
  - Does your code conform to our coding guidelines?
  - Did you write sufficient test code?
  - Did you write sufficient documentation?
- Also, take a look at [our compatibility policy](https://docs.chainer.org/en/stable/compatibility.html).

### CODE REVIEW

**Comment by unnonouno** [comment]
jenkins test this please

File: chainer/functions/loss/ctc.py [hunk 1]
```diff
@@ -405,5 +473,43 @@ def connectionist_temporal_classification(
...
+            raise ValueError("For cuDNN CTC, you must use 'blank_symbol' = 0")
+
+        # For cuDNN CTC, you can set `deterministic`.
+        deterministic = configuration.config.cudnn_deterministic
+        if isinstance(label_length, chainer.Variable):
+            input_length = input_length.data
+        if isinstance(label_length, chainer.Variable):
+            label_length = label_length.data
+
+        if input_length.max() != len(x):
```
[the remark this answers is not in the document]
Comment (unnonouno) [reply chainer/functions/loss/ctc.py]: Do you have a test for this case?

**Comment by unnonouno** [comment]
Actually we don't need to treat integer arrays as input because they are not backpropable. You can set integer arrays in `__init__`.

**Comment by aonotas** [comment]
Thank you for review!

I didn't add the tests for 255 lengths and 256 lengths so I will add these tests later.

I feel `xp is not numpy` is require to detect gpu or cpu.
https://github.com/chainer/chainer/blob/v5/chainer/functions/connection/n_step_lstm.py#L425
(when I remove `xp is not numpy`, cpu test is not passed.)


> Actually we don't need to treat integer arrays as input because they are not backpropable. You can set integer arrays in __init__.

Should I change the interface of CudnnCTC?
(should I change __init__ and __call__ in CudnnCTC?)

**Comment by unnonouno** [comment]
I mean "move this check code outside of `if xp is not numpy and use_cudnn:`"

**Comment by unnonouno** [comment]
Yes, please change interface of `__init__` and `__call__`.

**Comment by aonotas** [comment]
I will change the interface, sorry for late work...

**Comment by okuta** [comment]
Jenkins, test this please.

File: chainer/functions/loss/ctc.py [hunk 2]
```diff
@@ -411,5 +482,52 @@ def connectionist_temporal_classification(
...
+
+    if use_cudnn and x.dtype != xp.float32:
+        warnings.warn("For cuDNN, you must use `xp.float32`", UserWarning)
+        use_cudnn = False
+
+    if use_cudnn and xp is not numpy:
+        # For cuDNN CTC, you can set `deterministic`.
+        deterministic = configuration.config.cudnn_deterministic
+        if isinstance(label_length, chainer.Variable):
+            input_length = input_length.data
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]:
```suggestion
            input_length = input_length.array
```

File: chainer/functions/loss/ctc.py [hunk 3]
```diff
@@ -411,5 +482,52 @@ def connectionist_temporal_classification(
...
+        if isinstance(label_length, chainer.Variable):
+            label_length = label_length.data
```
Comment (okuta) [inline chainer/functions/loss/ctc.py]:
```suggestion
            label_length = label_length.array
```

**Comment by aonotas** [comment]
I fixed my code. 
Could you test it?

**Comment by okuta** [comment]
Jenkins, test this please.

**Pull request closed without merging**